File: scripts/smoke-backend.py
#!/usr/bin/env python3
"""Bounded Compose smoke runner for the documented admin and RTMP contracts.

The runner deliberately treats all server payloads as private input: it uses the
values required to drive the smoke flow but never writes response data,
credentials, cookies, CSRF tokens, stream names, or RTMP keys to stdout/stderr.
"""

from __future__ import annotations

import argparse
import http.cookiejar
import json
import math
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
import xml.etree.ElementTree as element_tree
from dataclasses import dataclass
from typing import Any, Callable


API_PREFIX = "/api/v0/admin"
SMOKE_TRACK_TITLE = "web10-compose-smoke"
REQUEST_TIMEOUT_SECONDS = 5.0
POLL_INTERVAL_SECONDS = 1.0
MAX_RESPONSE_BYTES = 2 * 1024 * 1024


@dataclass(frozen=True)
class Deadline:
    """A monotonic deadline shared by a complete smoke mode."""

    end_monotonic: float

    @classmethod
    def from_timeout(cls, timeout_seconds: float) -> "Deadline":
        return cls(time.monotonic() + timeout_seconds)

    def remaining(self) -> float:
        return max(0.0, self.end_monotonic - time.monotonic())

    def request_timeout(self) -> float:
        remaining = self.remaining()
        if remaining <= 0.0:
            raise OperationError("deadline", "timeout")
        return min(REQUEST_TIMEOUT_SECONDS, remaining)

    def sleep_until_next_poll(self) -> None:
        remaining = self.remaining()
        if remaining > 0.0:
            time.sleep(min(POLL_INTERVAL_SECONDS, remaining))


class OperationError(Exception):
    """A deliberately redacted failure: callers may report only these fields."""

    def __init__(self, operation: str, status: str) -> None:
        super().__init__(operation, status)
        self.operation = operation
        self.status = status


@dataclass(frozen=True)
class Response:
    status: int | None
    body: bytes | None

    @property
    def status_text(self) -> str:
        return "unavailable" if self.status is None else str(self.status)


class AdminHttpClient:
    """Small stdlib-only client for the cookie and synchronizer-token contract."""

    def __init__(self, base_url: str, rtmp_stat_url: str) -> None:
        self._base_url = validate_base_url(base_url, "base-url")
        self._rtmp_stat_url = validate_http_url(rtmp_stat_url, "rtmp-stat-url")
        self._cookies = http.cookiejar.CookieJar()
        self._opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(self._cookies))
        self._csrf_token: str | None = None

    def login(self, username: str, password: str, deadline: Deadline) -> None:
        response = self._request_json(
            "login",
            "POST",
            "/api/v0/admin/auth/login",
            {"username": username, "password": password},
            deadline,
            include_csrf=False,
        )
        payload = require_json_object("login", response, 200)
        csrf_token = payload.get("csrfToken")
        if not isinstance(csrf_token, str) or not csrf_token:
            raise OperationError("login", "invalid-response")
        if not self._has_admin_session_cookie():
            raise OperationError("login", "invalid-response")
        self._csrf_token = csrf_token

    def get_admin(self, operation: str, path: str, deadline: Deadline) -> Response:
        return self._request("GET", operation, path, None, deadline, include_csrf=False)

    def post_admin(self, operation: str, path: str, payload: dict[str, Any], deadline: Deadline) -> Response:
        return self._request_json(operation, "POST", path, payload, deadline, include_csrf=True)

    def get_rtmp_stat(self, deadline: Deadline) -> Response:
        return self._request_url("GET", "rtmp-stat", self._rtmp_stat_url, None, deadline, {})

    def _request_json(
        self,
        operation: str,
        method: str,
        path: str,
        payload: dict[str, Any],
        deadline: Deadline,
        *,
        include_csrf: bool,
    ) -> Response:
        return self._request(method, operation, path, payload, deadline, include_csrf=include_csrf)

    def _request(
        self,
        method: str,
        operation: str,
        path: str,
        payload: dict[str, Any] | None,
        deadline: Deadline,
        *,
        include_csrf: bool,
    ) -> Response:
        if not path.startswith("/"):
            raise OperationError(operation, "invalid-path")
        headers: dict[str, str] = {"Accept": "application/json", "User-Agent": "web10-backend-smoke"}
        if include_csrf:
            if self._csrf_token is None:
                raise OperationError(operation, "not-authenticated")
            headers["X-CSRF-Token"] = self._csrf_token
        return self._request_url(method, operation, self._base_url + path, payload, deadline, headers)

    def _request_url(
        self,
        method: str,
        operation: str,
        url: str,
        payload: dict[str, Any] | None,
        deadline: Deadline,
        headers: dict[str, str],
    ) -> Response:
        data: bytes | None = None
        request_headers = dict(headers)
        if payload is not None:
            data = json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
            request_headers["Content-Type"] = "application/json; charset=utf-8"

        request = urllib.request.Request(url, data=data, headers=request_headers, method=method)
        try:
            with self._opener.open(request, timeout=deadline.request_timeout()) as response:
                return Response(response.status, read_limited(response))
        except urllib.error.HTTPError as error:
            # Do not read error bodies: problem-details payloads are intentionally never displayed.
            error.close()
            return Response(error.code, None)
        except (OSError, TimeoutError, urllib.error.URLError):
            return Response(None, None)

    def _has_admin_session_cookie(self) -> bool:
        return any(cookie.name == "web10_admin_session" and bool(cookie.value) for cookie in self._cookies)


class SmokeRunner:
    def __init__(self, client: AdminHttpClient, deadline: Deadline) -> None:
        self._client = client
        self._deadline = deadline

    def authenticate(self, username: str, password: str) -> None:
        self._client.login(username, password, self._deadline)

    def run_live(self) -> None:
        track_id = self._scan_and_find_track()
        self._enqueue_track(track_id)
        self._wait_for_live_rtmp()

    def run_restart_live(self) -> None:
        self._stream_control("restart")
        track_id = self._find_smoke_track()
        self._enqueue_track(track_id)
        self._wait_for_live_rtmp()

    def run_expect_output_failure(self) -> None:
        def output_failed() -> tuple[bool, str | None]:
            snapshot, http_status = self._get_stream_status()
            if snapshot is None:
                return False, http_status
            accepted_statuses = {"degraded", "restarting", "failed"}
            return (
                snapshot.get("status") in accepted_statuses
                and snapshot.get("failureReason") == "RTMP output failed",
                None,
            )

        self._wait_for("output-failure", output_failed)

    def run_recover(self) -> None:
        first_restart = self._stream_control("restart")
        track_id = self._find_smoke_track()
        self._enqueue_track(track_id)
        self._wait_for_live_rtmp()

        stopped = self._stream_control("stop")
        self._wait_for_offline("stream-stop", required_generation=stopped, require_stopped=True)

        second_restart = self._stream_control("restart")
        if second_restart <= first_restart:
            raise OperationError("stream-restart", "invalid-response")
        self._wait_for_resumed(second_restart)

    def run_expect_offline(self) -> None:
        self._wait_for_offline("offline")

    def _scan_and_find_track(self) -> str:
        response = self._client.post_admin("library-scan", f"{API_PREFIX}/library/scan", {}, self._deadline)
        payload = require_json_object("library-scan", response, 202)
        scan_job_id = require_nonempty_string("library-scan", payload, "scanJobId")
        self._wait_for_scan(scan_job_id)
        return self._find_smoke_track()

    def _wait_for_scan(self, scan_job_id: str) -> None:
        path = f"{API_PREFIX}/library/scan/{urllib.parse.quote(scan_job_id, safe='')}"

        def scan_completed() -> tuple[bool, str | None]:
            response = self._client.get_admin("library-scan", path, self._deadline)
            if response.status != 200:
                return False, response.status_text
            payload = require_json_object("library-scan", response, 200)
            scan_status = payload.get("status")
            if scan_status == "failed":
                raise OperationError("library-scan", "failed")
            if not isinstance(scan_status, str):
                raise OperationError("library-scan", "invalid-response")
            return scan_status == "completed", None

        self._wait_for("library-scan", scan_completed)

    def _find_smoke_track(self) -> str:
        query = urllib.parse.urlencode({"query": SMOKE_TRACK_TITLE, "limit": "100"})
        response = self._client.get_admin("track-lookup", f"{API_PREFIX}/tracks?{query}", self._deadline)
        if response.status != 200:
            raise OperationError("track-lookup", response.status_text)
        tracks = require_json_array("track-lookup", response, 200)
        matches: list[dict[str, Any]] = []
        for candidate in tracks:
            if isinstance(candidate, dict) and candidate.get("title") == SMOKE_TRACK_TITLE:
                matches.append(candidate)

        if not matches:
            raise OperationError("track-lookup", "not-found")
        if len(matches) != 1:
            raise OperationError("track-lookup", "ambiguous")

        track_id = matches[0].get("id")
        if not isinstance(track_id, str) or not track_id or matches[0].get("hasCachedFile") is not True:
            raise OperationError("track-lookup", "invalid-response")
        return track_id

    def _enqueue_track(self, track_id: str) -> None:
        response = self._client.post_admin(
            "playback-queue",
            f"{API_PREFIX}/playback/queue",
            {"trackId": track_id},
            self._deadline,
        )
        payload = require_json_object("playback-queue", response, 202)
        require_nonempty_string("playback-queue", payload, "queueItemId")

    def _stream_control(self, action: str) -> int:
        response = self._client.post_admin(
            f"stream-{action}",
            f"{API_PREFIX}/stream-node/{action}",
            {},
            self._deadline,
        )
        payload = require_json_object(f"stream-{action}", response, 202)
        generation = payload.get("restartGeneration")
        if isinstance(generation, bool) or not isinstance(generation, int) or generation < 0:
            raise OperationError(f"stream-{action}", "invalid-response")
        return generation

    def _get_stream_status(self) -> tuple[dict[str, Any] | None, str | None]:
        response = self._client.get_admin("stream-status", f"{API_PREFIX}/stream-node/status", self._deadline)
        if response.status != 200:
            return None, response.status_text
        return require_json_object("stream-status", response, 200), None

    def _wait_for_live_rtmp(self) -> None:
        def live_rtmp() -> tuple[bool, str | None]:
            stream_status, stream_http_status = self._get_stream_status()
            if stream_status is None:
                return False, stream_http_status

            heartbeat_is_live = (
                stream_status.get("status") == "live"
                and stream_status.get("bitrateKbps") == 192
                and isinstance(stream_status.get("lastHeartbeatUtc"), str)
                and bool(stream_status["lastHeartbeatUtc"])
                and stream_status.get("failureReason") is None
            )
            if not heartbeat_is_live:
                return False, None

            rtmp_response = self._client.get_rtmp_stat(self._deadline)
            if rtmp_response.status != 200 or rtmp_response.body is None:
                return False, rtmp_response.status_text
            return rtmp_stat_has_active_stream(rtmp_response.body), None

        self._wait_for("live", live_rtmp)

    def _wait_for_offline(
        self,
        operation: str,
        required_generation: int | None = None,
        require_stopped: bool = False,
    ) -> None:
        def offline() -> tuple[bool, str | None]:
            snapshot, http_status = self._get_stream_status()
            if snapshot is None:
                return False, http_status
            if required_generation is not None and snapshot.get("restartGeneration") != required_generation:
                return False, None
            return (
                snapshot.get("status") == "offline"
                and (not require_stopped or snapshot.get("desiredState") == "stopped"),
                None,
            )

        self._wait_for(operation, offline)

    def _wait_for_resumed(self, required_generation: int) -> None:
        def resumed() -> tuple[bool, str | None]:
            snapshot, http_status = self._get_stream_status()
            if snapshot is None:
                return False, http_status
            generation = snapshot.get("restartGeneration")
            if isinstance(generation, bool) or not isinstance(generation, int):
                raise OperationError("stream-restart", "invalid-response")
            return (
                generation >= required_generation
                and snapshot.get("desiredState") == "running"
                and snapshot.get("status") in {"starting", "live"},
                None,
            )

        self._wait_for("stream-restart", resumed)

    def _wait_for(self, operation: str, probe: Callable[[], tuple[bool, str | None]]) -> None:
        last_http_status: str | None = None
        while self._deadline.remaining() > 0.0:
            complete, observed_http_status = probe()
            if complete:
                return
            if observed_http_status is not None:
                last_http_status = observed_http_status
            self._deadline.sleep_until_next_poll()
        raise OperationError(operation, last_http_status or "timeout")


def read_limited(response: Any) -> bytes:
    body = response.read(MAX_RESPONSE_BYTES + 1)
    if len(body) > MAX_RESPONSE_BYTES:
        return b""
    return body


def decode_json(operation: str, response: Response, expected_status: int) -> Any:
    if response.status != expected_status:
        raise OperationError(operation, response.status_text)
    if response.body is None or response.body == b"":
        raise OperationError(operation, "invalid-response")
    try:
        return json.loads(response.body.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        raise OperationError(operation, "invalid-response") from None


def require_json_object(operation: str, response: Response, expected_status: int) -> dict[str, Any]:
    payload = decode_json(operation, response, expected_status)
    if not isinstance(payload, dict):
        raise OperationError(operation, "invalid-response")
    return payload


def require_json_array(operation: str, response: Response, expected_status: int) -> list[Any]:
    payload = decode_json(operation, response, expected_status)
    if not isinstance(payload, list):
        raise OperationError(operation, "invalid-response")
    return payload


def require_nonempty_string(operation: str, payload: dict[str, Any], field: str) -> str:
    value = payload.get(field)
    if not isinstance(value, str) or not value:
        raise OperationError(operation, "invalid-response")
    return value


def rtmp_stat_has_active_stream(document: bytes) -> bool:
    try:
        root = element_tree.fromstring(document)
    except element_tree.ParseError:
        raise OperationError("rtmp-stat", "invalid-response") from None

    # Debian's nginx-rtmp 1.1.4 reports an active publisher through its root
    # connection counters, but does not emit a <stream> node when no viewers
    # are attached. Both values are session-scoped and reset to zero after the
    # publisher disconnects, so they prove an active ingress without logging
    # the private stream key.
    counters = {
        local_xml_name(element.tag): (element.text or "").strip()
        for element in root
        if local_xml_name(element.tag) in {"naccepted", "bytes_in"}
    }
    try:
        return int(counters.get("naccepted", "0")) > 0 and int(counters.get("bytes_in", "0")) > 0
    except ValueError:
        return False


def local_xml_name(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def validate_http_url(value: str, argument: str) -> str:
    try:
        parsed = urllib.parse.urlsplit(value)
    except ValueError:
        raise OperationError("arguments", "invalid") from None
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise OperationError("arguments", "invalid")
    if parsed.username is not None or parsed.password is not None:
        raise OperationError("arguments", "invalid")
    return value


def validate_base_url(value: str, argument: str) -> str:
    normalized = validate_http_url(value, argument)
    parsed = urllib.parse.urlsplit(normalized)
    if parsed.path not in {"", "/"} or parsed.query or parsed.fragment:
        raise OperationError("arguments", "invalid")
    return normalized.rstrip("/")


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Run the bounded backend/RTMP Compose smoke modes without displaying private payload data."
    )
    parser.add_argument(
        "--mode",
        choices=("live", "restart-live", "expect-output-failure", "recover", "expect-offline"),
        required=True,
        help="Smoke phase to execute.",
    )
    parser.add_argument("--base-url", required=True, help="API origin, for example http://localhost:8080.")
    parser.add_argument("--rtmp-stat-url", required=True, help="Local RTMP sink status endpoint.")
    parser.add_argument("--username", required=True, help="Admin username used only for the login request.")
    parser.add_argument("--password", required=True, help="Admin password used only for the login request.")
    parser.add_argument(
        "--timeout-seconds",
        type=float,
        default=60.0,
        metavar="SECONDS",
        help="Monotonic deadline for the complete mode (default: 60).",
    )
    return parser


def validate_arguments(args: argparse.Namespace) -> None:
    if not math.isfinite(args.timeout_seconds) or args.timeout_seconds <= 0.0:
        raise OperationError("arguments", "invalid")
    if not isinstance(args.username, str) or not isinstance(args.password, str):
        raise OperationError("arguments", "invalid")


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    try:
        validate_arguments(args)
        deadline = Deadline.from_timeout(args.timeout_seconds)
        client = AdminHttpClient(args.base_url, args.rtmp_stat_url)
        runner = SmokeRunner(client, deadline)
        runner.authenticate(args.username, args.password)

        if args.mode == "live":
            runner.run_live()
        elif args.mode == "restart-live":
            runner.run_restart_live()
        elif args.mode == "expect-output-failure":
            runner.run_expect_output_failure()
        elif args.mode == "recover":
            runner.run_recover()
        else:
            runner.run_expect_offline()
    except OperationError as failure:
        print(f"{failure.operation} status={failure.status}", file=sys.stderr)
        return 1
    except KeyboardInterrupt:
        print("main status=interrupted", file=sys.stderr)
        return 130
    except Exception:
        # Keep unexpected runtime failures redacted just like HTTP failures.
        print("main status=failed", file=sys.stderr)
        return 1

    print(f"{args.mode} status=passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/stream-node/Web10.Radio.StreamNode.Tools/SmokeBackend.fs
namespace Web10.Radio.StreamNode.Tools

open System
open System.Globalization
open System.Linq
open System.Text.Json
open System.Threading
open System.Threading.Tasks
open System.Xml.Linq

[<CLIMutable>]
type SmokeOptions =
    { Mode: string
      BaseUrl: string
      RtmpStatUrl: string
      Username: string
      Password: string
      TimeoutSeconds: float }

module SmokeBackend =
    let private supportedModes =
        Set.ofList [ "restart-live"; "flac-cue"; "reorder"; "skip"; "restart-current"; "play-now"; "expect-output-failure"; "recover" ]

    let options (arguments: Map<string, string>) =
        let mode = Parsing.option "mode" "" arguments
        if not (Set.contains mode supportedModes) then
            raise (ToolError("mode", "invalid"))
        let baseUrl = Parsing.required "base-url" arguments
        let rtmpStatUrl = Parsing.required "rtmp-stat-url" arguments
        let username = Parsing.required "username" arguments
        let password = Parsing.required "password" arguments
        let timeoutSeconds = Parsing.option "timeout-seconds" "60" arguments |> Parsing.positiveFloat "timeout-seconds"
        { Mode = mode
          BaseUrl = baseUrl.TrimEnd('/')
          RtmpStatUrl = rtmpStatUrl
          Username = username
          Password = password
          TimeoutSeconds = timeoutSeconds }

    let private property (name: string) (document: JsonElement) =
        let mutable value = Unchecked.defaultof<JsonElement>
        if document.TryGetProperty(name, &value) then Some value else None

    let private stringValue (name: string) (document: JsonElement) = Document.string name document

    let private int64Value (name: string) (document: JsonElement) =
        match property name document with
        | Some value ->
            let mutable number = 0L
            if value.TryGetInt64(&number) then Some number else None
        | None -> None

    let private statusSnapshot (admin: AdminConnection) (deadline: Time.Deadline) cancellationToken =
        task {
            let! response = admin.Get("stream-status", "/api/v0/admin/stream-node/status", deadline, cancellationToken)
            if response.Status <> 200 then return None
            else
                match response.Body |> Option.bind Json.tryDocument with
                | Some document -> return Some document
                | None -> return None
        }

    let private restart (admin: AdminConnection) (deadline: Time.Deadline) cancellationToken =
        task {
            let! response = admin.PostWithCsrf("stream-restart", "/api/v0/admin/stream-node/restart", Map.empty<string, obj>, 202, deadline, cancellationToken)
            match response.Body |> Option.bind Json.tryDocument with
            | Some document ->
                match int64Value "restartGeneration" document with
                | Some generation when generation >= 0L -> return generation
                | _ -> return raise (ToolError("stream-restart", "invalid-response"))
            | None -> return raise (ToolError("stream-restart", "invalid-response"))
        }

    let private stop (admin: AdminConnection) (deadline: Time.Deadline) cancellationToken =
        task {
            let! response = admin.PostWithCsrf("stream-stop", "/api/v0/admin/stream-node/stop", Map.empty<string, obj>, 202, deadline, cancellationToken)
            match response.Body |> Option.bind Json.tryDocument with
            | Some document ->
                match int64Value "restartGeneration" document with
                | Some generation when generation >= 0L -> return generation
                | _ -> return raise (ToolError("stream-stop", "invalid-response"))
            | None -> return raise (ToolError("stream-stop", "invalid-response"))
        }

    let private findTrack (expectedTitle: string) (admin: AdminConnection) (deadline: Time.Deadline) cancellationToken =
        task {
            let query = Uri.EscapeDataString expectedTitle
            let! response = admin.Get("track-lookup", "/api/v0/admin/tracks?query=" + query + "&limit=100", deadline, cancellationToken)
            if response.Status <> 200 then
                return raise (ToolError("track-lookup", string response.Status))
            else
                let document = response.Body |> Option.bind Json.tryDocument |> Option.defaultWith (fun () -> raise (ToolError("track-lookup", "invalid-response")))
                let candidates = Document.trackCandidates document |> Option.defaultValue [||]
                match candidates |> Array.filter (fun track -> track.Title = expectedTitle && track.HasCachedFile) with
                | [| candidate |] -> return candidate.Id
                | [||] -> return raise (ToolError("track-lookup", "not-found"))
                | _ -> return raise (ToolError("track-lookup", "ambiguous"))
        }

    let private enqueue (admin: AdminConnection) trackId (deadline: Time.Deadline) cancellationToken =
        task {
            let! response = admin.PostWithCsrf("playback-queue", "/api/v0/admin/playback/queue", {| TrackId = trackId |}, 202, deadline, cancellationToken)
            let document = response.Body |> Option.bind Json.tryDocument |> Option.defaultWith (fun () -> raise (ToolError("playback-queue", "invalid-response")))
            let _ = Document.stringField "queueItemId" document
            return ()
        }

    let private getStringField name operation document =
        match stringValue name document with
        | Some value -> value
        | None -> raise (ToolError(operation, "invalid-response"))

    let private waitFor operation (deadline: Time.Deadline) cancellationToken (probe: unit -> Task<bool * string option>) =
        task {
            let mutable lastStatus: string option = None
            let mutable complete = false
            while not complete && Time.remaining deadline > TimeSpan.Zero do
                let! success, observed = probe ()
                if success then complete <- true
                else
                    match observed with
                    | Some value -> lastStatus <- Some value
                    | None -> ()
                    do! Time.delay deadline.Provider (TimeSpan.FromSeconds 1.0) cancellationToken
            if not complete then raise (ToolError(operation, lastStatus |> Option.defaultValue "timeout"))
        }

    let private waitForLive admin rtmpStatUrl (deadline: Time.Deadline) cancellationToken =
        waitFor "live" deadline cancellationToken (fun () ->
            task {
                let! status = statusSnapshot admin deadline cancellationToken
                match status with
                | None -> return false, Some "unavailable"
                | Some document ->
                    let isLive = stringValue "status" document = Some "live" && (stringValue "failureReason" document).IsNone
                    if not isLive then
                        return false, None
                    else
                        let! response = admin.GetRtmpStat(rtmpStatUrl, deadline, cancellationToken)
                        if response.Status <> 200 then
                            return false, Some(string response.Status)
                        else
                            let active =
                                match response.Body with
                                | None -> false
                                | Some body ->
                                    try
                                        let xml = XDocument.Parse body
                                        let values =
                                            xml.Descendants()
                                            |> Seq.filter (fun node -> node.Name.LocalName = "naccepted" || node.Name.LocalName = "bytes_in")
                                            |> Seq.map (fun node -> node.Name.LocalName, node.Value.Trim())
                                            |> Map.ofSeq
                                        match Map.tryFind "naccepted" values, Map.tryFind "bytes_in" values with
                                        | Some accepted, Some bytesIn -> Int64.Parse(accepted, CultureInfo.InvariantCulture) > 0L && Int64.Parse(bytesIn, CultureInfo.InvariantCulture) > 0L
                                        | _ -> false
                                    with _ -> false
                            return active, None
            })

    let private waitForNowPlaying (admin: AdminConnection) (expectedTrackId: string) (deadline: Time.Deadline) cancellationToken =
        waitFor "cue-now-playing" deadline cancellationToken (fun () ->
            task {
                let! state = admin.GetJson("player-state", "/api/v0/player/state", deadline, cancellationToken)
                match property "nowPlaying" state with
                | Some nowPlaying when nowPlaying.ValueKind = JsonValueKind.Object ->
                    return stringValue "trackId" nowPlaying = Some expectedTrackId, stringValue "trackId" nowPlaying
                | _ -> return false, Some "missing-now-playing"
            })

    let private waitForOffline admin requiredGeneration requireStopped (deadline: Time.Deadline) cancellationToken =
        waitFor "offline" deadline cancellationToken (fun () ->
            task {
                let! status = statusSnapshot admin deadline cancellationToken
                match status with
                | None -> return false, Some "unavailable"
                | Some document ->
                    let generationMatches = requiredGeneration |> Option.forall (fun expected -> int64Value "restartGeneration" document = Some expected)
                    let stopped = not requireStopped || stringValue "desiredState" document = Some "stopped"
                    return generationMatches && stopped && stringValue "status" document = Some "offline", None
            })

    let private waitForResumed admin requiredGeneration (deadline: Time.Deadline) cancellationToken =
        waitFor "stream-restart" deadline cancellationToken (fun () ->
            task {
                let! status = statusSnapshot admin deadline cancellationToken
                match status with
                | None -> return false, Some "unavailable"
                | Some document ->
                    let generation = int64Value "restartGeneration" document |> Option.defaultValue -1L
                    return generation >= requiredGeneration && stringValue "desiredState" document = Some "running" && (stringValue "status" document = Some "starting" || stringValue "status" document = Some "live"), None
            })

    let private queueIds (document: JsonElement) =
        match property "queue" document with
        | Some queue ->
            match property "items" queue with
            | Some items when items.ValueKind = JsonValueKind.Array ->
                items.EnumerateArray()
                |> Seq.choose (fun item ->
                    match stringValue "status" item with
                    | Some "queued" -> stringValue "id" item |> Option.orElse (stringValue "queueItemId" item)
                    | _ -> None)
                |> Seq.toArray
            | _ -> [||]
        | _ -> [||]

    let private reorder (admin: AdminConnection) (deadline: Time.Deadline) cancellationToken =
        task {
            let! state = admin.GetJson("queue-state", "/api/v0/player/state", deadline, cancellationToken)
            let ids = queueIds state
            if ids.Length > 0 then
                let reversed = ids |> Array.rev
                let! _ = admin.PutWithCsrf("playback-reorder", "/api/v0/admin/playback/queue/order", {| QueueItemIds = reversed |}, 200, deadline, cancellationToken)
                return ()
            else
                return ()
        }

    let private control (admin: AdminConnection) (route: string) (operation: string) (deadline: Time.Deadline) cancellationToken =
        task {
            let! _ = admin.PostWithCsrf(operation, route, Map.empty<string, obj>, 202, deadline, cancellationToken)
            return ()
        }

    let run (provider: TimeProvider) (settings: SmokeOptions) cancellationToken =
        task {
            let deadline = Time.create provider (TimeSpan.FromSeconds settings.TimeoutSeconds)
            let admin = new AdminConnection(settings.BaseUrl, settings.Username, settings.Password)
            do! admin.Login(deadline, cancellationToken)
            match settings.Mode with
            | "restart-live" ->
                let! _ = restart admin deadline cancellationToken
                let! track = findTrack "web10-compose-smoke" admin deadline cancellationToken
                do! enqueue admin track deadline cancellationToken
                do! waitForLive admin settings.RtmpStatUrl deadline cancellationToken
            | "flac-cue" ->
                let! _ = restart admin deadline cancellationToken
                let! track = findTrack "web10-compose-cue-track-02" admin deadline cancellationToken
                let! _ = admin.PostWithCsrf("cue-play-now", "/api/v0/admin/playback/play-now", {| TrackId = track |}, 202, deadline, cancellationToken)
                do! waitForNowPlaying admin track deadline cancellationToken
                do! waitForLive admin settings.RtmpStatUrl deadline cancellationToken
            | "reorder" -> do! reorder admin deadline cancellationToken
            | "skip" -> do! control admin "/api/v0/admin/playback/skip" "playback-skip" deadline cancellationToken
            | "restart-current" -> do! control admin "/api/v0/admin/playback/restart-current" "playback-restart-current" deadline cancellationToken
            | "play-now" ->
                let! track = findTrack "web10-compose-smoke" admin deadline cancellationToken
                let! _ = admin.PostWithCsrf("play-now", "/api/v0/admin/playback/play-now", {| TrackId = track |}, 202, deadline, cancellationToken)
                return ()
            | "expect-output-failure" ->
                do!
                    waitFor "output-failure" deadline cancellationToken (fun () ->
                        task {
                            let! status = statusSnapshot admin deadline cancellationToken
                            match status with
                            | Some document ->
                                let state = stringValue "status" document
                                let reason = stringValue "failureReason" document
                                return (state = Some "degraded" || state = Some "restarting" || state = Some "failed") && reason = Some "RTMP output failed", None
                            | None -> return false, Some "unavailable"
                        })
            | "recover" ->
                let! firstGeneration = restart admin deadline cancellationToken
                let! track = findTrack "web10-compose-smoke" admin deadline cancellationToken
                do! enqueue admin track deadline cancellationToken
                do! waitForLive admin settings.RtmpStatUrl deadline cancellationToken
                let! stoppedGeneration = stop admin deadline cancellationToken
                do! waitForOffline admin (Some stoppedGeneration) true deadline cancellationToken
                let! secondGeneration = restart admin deadline cancellationToken
                if secondGeneration <= firstGeneration then raise (ToolError("stream-restart", "invalid-response"))
                do! waitForResumed admin secondGeneration deadline cancellationToken
            | _ -> raise (ToolError("mode", "invalid"))
        }
